cupy/cudnn.pyx

```diff
@@ -449,3 +465,315 @@ class DropoutStates(object):
+
+
+cpdef tuple _get_algorithm_fwd(
+        core.ndarray x, core.ndarray W, core.ndarray y, tuple conv_param,
+        size_t handle, size_t x_desc, size_t filter_desc, size_t conv_desc,
+        size_t y_desc, size_t max_workspace_size):
+    key = (x.data.device.id, x.shape, W.shape, y.shape, conv_param,
+           max_workspace_size)
+    if key in _algorithm_fwd:
+        return _algorithm_fwd[key]
+    workspace = memory.alloc(max_workspace_size)
+    ret = cudnn.findConvolutionForwardAlgorithmEx(
+        handle, x_desc, x.data.ptr, filter_desc, W.data.ptr, conv_desc, y_desc,
+        y.data.ptr, 1, workspace.ptr, max_workspace_size)
+    algo = (ret[0]['algo'], ret[0]['memory'])
+    _algorithm_fwd[key] = algo
+    return algo
+
+
+cpdef tuple _get_algorithm_bwd_filter(
+        core.ndarray x, core.ndarray dy, core.ndarray dW, tuple conv_param,
+        size_t handle, size_t x_desc, size_t dy_desc, size_t conv_desc,
+        size_t filter_desc, size_t max_workspace_size):
+    key = (x.data.device.id, x.shape, dW.shape, dy.shape, conv_param,
+           max_workspace_size)
+    if key in _algorithm_bwd_filter:
+        return _algorithm_bwd_filter[key]
+    workspace = memory.alloc(max_workspace_size)
+    ret = cudnn.findConvolutionBackwardFilterAlgorithmEx(
+        handle, x_desc, x.data.ptr, dy_desc, dy.data.ptr, conv_desc,
+        filter_desc, dW.data.ptr, 1, workspace.ptr, max_workspace_size)
+    algo = (ret[0]['algo'], ret[0]['memory'])
+    _algorithm_bwd_filter[key] = algo
+    return algo
+
+
+cpdef tuple _get_algorithm_bwd_data(
+        core.ndarray W, core.ndarray x, core.ndarray y, tuple conv_param,
+        size_t handle, size_t filter_desc, size_t x_desc, size_t conv_desc,
+        size_t y_desc, size_t max_workspace_size):
+    key = (x.data.device.id, W.shape, x.shape, y.shape, conv_param,
+           max_workspace_size)
+    if key in _algorithm_bwd_data:
+        return _algorithm_bwd_data[key]
+    workspace = memory.alloc(max_workspace_size)
+    ret = cudnn.findConvolutionBackwardDataAlgorithmEx(
+        handle, filter_desc, W.data.ptr, x_desc, x.data.ptr,
+        conv_desc, y_desc, y.data.ptr, 1, workspace.ptr, max_workspace_size)
+    algo = (ret[0]['algo'], ret[0]['memory'])
+    _algorithm_bwd_data[key] = algo
+    return algo
+
+
+cpdef bint _should_use_tensor_core(
+        str tensor_code_mode, object dtype) except *:
+    if tensor_code_mode == 'auto':
+        use_tensor_core = is_tensor_core_available(dtype)
+    elif tensor_code_mode == 'always':
+        # TODO(oktua): more strict condition
+        use_tensor_core = _cudnn_version >= 7000
+    elif tensor_code_mode == 'never':
+        use_tensor_core = False
+    else:
+        raise ValueError(
+            'tensor_code_mode must be either of "always", "auto", or "never".')
+
+
+def convolution_forward(
+        core.ndarray x, core.ndarray W, core.ndarray b, core.ndarray y,
+        tuple pad, tuple stride, tuple dilation, int group,
```

Comment:
Should I change `group` to `groups` ? 
Please see [this](https://github.com/chainer/chainer/issues/4336).